primitives/src/circuit/merkle_tree/mod.rs

```diff
@@ -69,16 +93,211 @@ where
+
+    /// Ensure that the position of each node of the path is correctly encoded
+    /// Used for testing purposes.
+    /// * `merkle_path` - list of node of an authentication path
+    /// * `pos` - position of the missing leaf
+    /// * `returns` - list of variables corresponding to the authentication path
+    fn constrain_membership_proof(
+        &mut self,
+        merkle_path: &Self::MerklePathEncoding,
+        pos: M::Index,
+    ) -> Result<Self::MembershipProofVar, CircuitError>;
+
+    /// Computes the merkle root based on some element placed at a leaf and a
+    /// merkle path.
+    /// * `elem` - variables corresponding to the uid and the element value
+    ///   (e.g.: record commitment).
+    /// * `path_vars` - variables corresponding to the Merkle path.
+    /// * `return` - variable corresponding to the root value of the Merkle
+    ///   tree.
+    fn compute_merkle_root(
+        &mut self,
+        elem: LeafVar,
+        path_vars: &Self::MembershipProofVar,
+    ) -> Result<Variable, CircuitError>;
+}
+
+#[derive(Debug, Clone)]
+/// Circuit variable for a node in the Merkle path.
+pub struct Merkle3AryNodeVar {
+    /// First sibling of the node.
+    pub sibling1: Variable,
+    /// Second sibling of the node.
+    pub sibling2: Variable,
+    /// Boolean variable indicating whether the node is a left child.
+    pub is_left_child: BoolVar,
+    /// Boolean variable indicating whether the node is a right child.
+    pub is_right_child: BoolVar,
+}
+
+/// Circuit variable for a leaf element.
+#[derive(Debug, Clone)]
+pub struct LeafVar {
+    /// Position of the leaf element in the MT. Serves as UID.
+    pub uid: Variable,
+    /// The value of the leaf element.
+    pub elem: Variable,
+}
+
+/// Intermediate representation for a node in the Merkle path, used for
+/// constructing `Merkle3AryNodeVar`. For constraining the membership proof to
+/// be correct, we need to encode each node in the merkle path s.t. it contains
+/// the following information:
+/// * the two siblings of the node,
+/// * whether the node is a left child, right child, or in the middle.
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub(crate) struct Merkle3AryNodeRepr<M: MerkleTreeScheme> {
+    sibling1: M::NodeValue,
+    sibling2: M::NodeValue,
+    is_left_child: bool,
+    is_right_child: bool,
+}
+
+impl<M: MerkleTreeScheme> Merkle3AryNodeRepr<M> {
+    fn new(
+        sibling1: M::NodeValue,
+        sibling2: M::NodeValue,
+        is_left_child: bool,
+        is_right_child: bool,
+    ) -> Self {
+        Merkle3AryNodeRepr {
+            sibling1,
+            sibling2,
+            is_left_child,
+            is_right_child,
+        }
+    }
+}
+
+/// Intermediate encoding for Merkle path, used for constructing
+/// `Merkle3AryMembershipProofVar`. Constains a list of nodes in its path.
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub(crate) struct MerklePathRepr<M: MerkleTreeScheme> {
+    pub(crate) nodes: Vec<Merkle3AryNodeRepr<M>>,
+}
```

Comment:
not very sure why we need this "seemingly intermediate" struct. why can't we only have `Merkle3AryMemebershipProofVar` ?

Reply:
Yeah okey, I've refactored the code slightly to not use `Repr` structs anymore, in 721f9fd129cc054e9c66cb0c145060432a1b55aa